flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadeImpl.java

```diff
@@ -0,0 +1,453 @@
+/** Default {@link XaFacade} implementation. */
+@NotThreadSafe
+@Internal
+class XaFacadeImpl implements XaFacade {
+
+    private static final Logger LOG = LoggerFactory.getLogger(XaFacadeImpl.class);
+    private static final Set<Integer> TRANSIENT_ERR_CODES =
+            new HashSet<>(Arrays.asList(XA_RBTRANSIENT, XAER_RMFAIL));
+    private static final Set<Integer> HEUR_ERR_CODES =
+            new HashSet<>(Arrays.asList(XA_HEURRB, XA_HEURCOM, XA_HEURHAZ, XA_HEURMIX));
+    private static final int MAX_RECOVER_CALLS = 100;
+
+    private final Supplier<XADataSource> dataSourceSupplier;
+    private final Integer timeoutSec;
+    private transient XAResource xaResource;
+    private transient Connection connection;
+    private transient XAConnection xaConnection;
+
+    /** @return a non-serializable instance. */
+    static XaFacadeImpl fromXaDataSource(XADataSource ds) {
+        return new XaFacadeImpl(() -> ds, empty());
+    }
+
+    XaFacadeImpl(Supplier<XADataSource> dataSourceSupplier, Optional<Integer> timeoutSec) {
+        this.dataSourceSupplier = Preconditions.checkNotNull(dataSourceSupplier);
+        this.timeoutSec = timeoutSec.orElse(null);
+    }
+
+    @Override
+    public void open() throws SQLException, XAException {
+        Preconditions.checkState(!isOpen(), "already connected");
+        XADataSource ds = dataSourceSupplier.get();
+        xaConnection = ds.getXAConnection();
+        xaResource = xaConnection.getXAResource();
+        if (timeoutSec != null) {
+            xaResource.setTransactionTimeout(timeoutSec);
+        }
+        connection = xaConnection.getConnection();
+        connection.setReadOnly(false);
+        connection.setAutoCommit(false);
+        Preconditions.checkState(!connection.getAutoCommit());
+    }
+
+    @Override
+    public void close() throws SQLException {
+        if (connection != null) {
+            connection.close();
+            connection = null;
+        }
+        if (xaConnection != null) {
+            xaConnection.close();
+            xaConnection = null;
+        }
+        xaResource = null;
+    }
+
+    @Override
+    public Connection getConnection() {
+        Preconditions.checkNotNull(connection);
+        return connection;
+    }
+
+    @Override
+    public boolean isConnectionValid() {
+        return isOpen();
+    }
+
+    @Override
+    public Connection getOrEstablishConnection() throws SQLException, ClassNotFoundException {
+        if (!isOpen()) {
+            try {
+                open();
+            } catch (XAException e) {
+                throw new SQLException(e);
+            }
+        }
+        return connection;
+    }
+
+    @Override
+    public void closeConnection() {
+        try {
+            close();
+        } catch (SQLException e) {
+            LOG.warn("Connection close failed.", e);
+        }
+    }
+
+    @Override
+    public Connection reestablishConnection() {
+        throw new UnsupportedOperationException();
```

Comment:
I'm afraid not: a potential ongoing XA transaction should be resumed (or at least aborted). 
This would require either passing `Xid` from the outside or storing current `Xid` in this class. I'd rather avoid both of these options because they either complicate the caller or duplicate `Xid` tracking.